### ISSUE TITLE
Gradient computations inside Trainer.predict yield zero

### ISSUE THREAD

**Issue by henrik-schopmans** [Lightning-AI/lightning#16838]
### Bug description

When calculating the gradient (using functorch.grad) with respect to the input of some function inside of `trainer.predict`, the gradients are zero (see variable `force`). When calling the model directly (see variable `force_compare`), the gradient calculation works correctly and is non-zero.

### How to reproduce the bug

```python
import torch
import torch.nn as nn
import pytorch_lightning as pl
import functorch
from pytorch_lightning import Trainer
from torch.utils.data import TensorDataset, DataLoader


class TestModel(pl.LightningModule):
    def __init__(self):
        super(TestModel, self).__init__()
        self.lin_layer = nn.Linear(10, 1)

    def compute_E(self, x):
        output = self.lin_layer(x)
        return output[0], output[0]

    def forward(self, x):
        grad, output = functorch.vmap(functorch.grad(self.compute_E, has_aux=True))(x)
        F_force = -1 * grad
        return output, F_force

    def predict_step(self, batch, batch_idx, dataloader_idx=0):
        return self(batch[0])


if __name__ == "__main__":
    trainer = Trainer(
        accelerator="gpu",
        devices=1,
    )

    my_model = TestModel()

    input_data = torch.rand((100, 10))
    dataset = TensorDataset(input_data)
    loader = DataLoader(dataset, batch_size=100)

    E, force = trainer.predict(my_model, dataloaders=loader)[0]
    E_compare, force_compare = my_model(input_data)
```

### Environment

* CUDA:
	- GPU:
		- NVIDIA GeForce RTX 3070 Laptop GPU
	- available:         True
	- version:           11.6
* Lightning:
	- lightning-utilities: 0.7.0
	- pytorch-lightning: 1.9.3
	- torch:             1.13.1+cu116
	- torchaudio:        0.13.1+cu116
	- torchmetrics:      0.11.1
	- torchvision:       0.14.1+cu116
* Packages:
	- aiohttp:           3.8.4
	- aiosignal:         1.3.1
	- async-timeout:     4.0.2
	- attrs:             22.2.0
	- certifi:           2022.12.7
	- charset-normalizer: 3.0.1
	- frozenlist:        1.3.3
	- fsspec:            2023.1.0
	- idna:              3.4
	- lightning-utilities: 0.7.0
	- multidict:         6.0.4
	- numpy:             1.24.2
	- packaging:         23.0
	- pillow:            9.4.0
	- pip:               23.0.1
	- pytorch-lightning: 1.9.3
	- pyyaml:            6.0
	- requests:          2.28.2
	- setuptools:        67.4.0
	- torch:             1.13.1+cu116
	- torchaudio:        0.13.1+cu116
	- torchmetrics:      0.11.1
	- torchvision:       0.14.1+cu116
	- tqdm:              4.64.1
	- typing-extensions: 4.5.0
	- urllib3:           1.26.14
	- wheel:             0.38.4
	- yarl:              1.8.2
* System:
	- OS:                Linux
	- architecture:
		- 64bit
		- ELF
	- processor:         x86_64
	- python:            3.8.0
	- version:           #202212290932~1671652965~22.04~452ea9d SMP PREEMPT_DYNAMIC Wed D

### More info

This is not simply an issue of `with torch.no_grad()` probably used by pytorch lightning internally when calling the model inside `trainer.predict()`. If I add `with torch.no_grad()` to the forward function myself, the problem still does not occur when calling the model directly.

I am not sure if this is an issue of pytorch lightning or functorch. However, it is definitely unexpected behavior.

**Comment by carmocca** [comment]
We use the `torch.inference_mode()` context manager by default, not `no_grad`. You can change it to `no_grad` by setting `Trainer(inference_mode=False)`.

Can you give this a shot, see if it makes a difference?

**Comment by henrik-schopmans** [comment]
Setting `inference_mode=False` does indeed solve the problem. Thank you!
I am just using this setting now in combination with `torch.no_grad()`, which works well with functorch.
I am not sure how to evaluate this behavior: Maybe one should add a comment about this in the docs?

**Comment by carmocca** [comment]
I don't know if this is expected or a bug. I suggest asking on the PyTorch forums. You'd think that `functorch.grad` would work for neither or both context managers.

Closing as this is not PL specific!
